dimod/include/dimod/quadratic_model.h

```diff
@@ -0,0 +1,686 @@
+// Copyright 2021 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+namespace dimod {
+
+/// Encode the domain of a variable.
+enum Vartype {
+    BINARY,  ///< Variables that are either 0 or 1.
+    SPIN,    ///< Variables that are either -1 or 1.
+    INTEGER  ///< Variables that are integer valued.
+};
+
+/**
+ * Used internally by QuadraticModelBase to sparsely encode the neighborhood of 
+ * a variable.
+ *
+ * Internally, Neighborhoods keep two vectors, one of neighbors and the other
+ * of biases. Neighborhoods are designed to make access more like a standard
+ * library map.
+ */
+template <class Bias, class Neighbor>
+class Neighborhood {
+ public:
+    /// The first template parameter (Bias).
+    using bias_type = Bias;
+
+    /// The type for variable indices
+    using index_type = std::ptrdiff_t;
```

Comment:
I do not think we need a separate type for indexing, I mean class Neighbor could be just index_type, we will not index to anything that is larger than Neighbor , the only thing that may be larger than what Neighbor type can hold is num_interactions, so having separate index_type and Neighbor is just for aesthetic reasons, and I do not have any argument against that if it is not hard to remember and be consistent with the types. 

One question,will it be easy for you to declare these functions which use index_type in cython ?

Reply:
Ok, maybe I will just drop `neighbor_type` altogether and just use `index_type` to simplify it. And yes, cython can handle it just fine.